cpp/src/utilities/cuda_utils.hpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 BlazingDB, Inc.
+ *     Copyright 2019 Eyal Rozenberg <eyalroz@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_UTILITIES_CUDA_UTIL_CUH_
+#define CUDF_UTILITIES_CUDA_UTIL_CUH_
+
+/**
+ * @file Utility code for host-side interaction with CUDA by cuDF API
+ * function implementations
+ */
+
+#include <cuda_runtime_api.h>
+
+namespace cudf {
+
+namespace util {
+
+namespace cuda {
+
+/**
+ * A RAII gadget for top-level CUDF API function implementations, which may
+ * create their own streams. This will ensure the stream is destroyed when the API function returns,
+ * without you having to explicitly destroy it - and also when you exit due to an error.
+ */
+struct scoped_stream {
+    cudaStream_t stream_ { nullptr };
+
+    scoped_stream() {
+        cudaStreamCreate(&stream_);
+        assert(stream_ != nullptr and "Failed creating a CUDA stream");
+    }
+    operator cudaStream_t() { return stream_; }
+    ~scoped_stream() {
+        if (not std::uncaught_exception()) {
+            cudaStreamSynchronize(stream_);
+        }
+        cudaStreamDestroy(stream_);
```

Comment:
```suggestion
        CUDA_TRY(cudaStreamDestroy(stream_));
```

Reply:
Done.

Reply:
Actually, on second thought, maybe we should not throw exceptions from a destructor, but be content with a best-effort to destroy the stream.

Reply:
Good point. Let's say the stream got errantly destroyed by someone who "borrowed" it's handle from this wrapper. Then the wrapper would get a cudaErrorInvalidResource or something like that. I don't think that's worth terminating the app over...

Reply:
@harrism : So how about just:
```
~scoped_stream() {
     if (not std::uncaught_exception()) { cudaStreamSynchronize(stream_); }
     cudaStreamDestroy(stream_);
}
?

Reply:
CUDA doesn't throw exceptions, so this isn't sufficient. I think you need:

```
~scoped_stream() {
  if (not std::uncaught_exception()) { 
    if (cudaSuccess == cudaStreamSynchronize(stream_); 
      cudaStreamDestroy(stream_);
}
```

Hmmm, maybe you don't want to synchronize the stream in the destructor at all. Shouldn't synchronization be explicit by the user? Otherwise you might get duplicate syncs (not a huge problem but still...).

Reply:
I know CUDA doesn't throw exceptions - but we do, and so does thrust.

About synchronization - in a perfect world, our API functions would get a stream, enqueue work on it, and not synchronize it nor destroy it. But right now, we create a stream, enqueue work, synchronize and destroy; the wrapper encapsulates existing behavior.

Reply:
It's fine to check for Uncaught exceptions. My point is that if the stream is invalid, don't call cudaStreamDestroy (because it more than likely means it was destroyed already out from under you).  Synchronizing on a destroyed stream is a recoverable error -- no need to pile on more errors.

Reply:
How would I decide whether the stream is invalid? I mean, the exception could be something unrelated to the stream, like a division by 0, or invalid argument, or not enough memory to allocate a scratch buffer and so on, so I definitely need to try and destroy the stream if I can.

Reply:
You are conflating exceptions and cuda errors. Ignoring exceptions for the moment, don't you agree that if cudaStreamSynchronize tells you that the stream is invalid, you shouldn't then try to destroy the invalid stream?

Reply:
@harrism : I agree, but I tend to believe I don't want to synchronize a stream when an exception has been thrown. How about if I "check validity" using `cudaPeekAtLastError()`?

Reply:
I give up.  The code I suggested does not synchronize when an exception has been thrown:

```
~scoped_stream() {
  if (not std::uncaught_exception()) { 
    if (cudaSuccess == cudaStreamSynchronize(stream_); 
      cudaStreamDestroy(stream_);
  }
}
```

Reply:
Going full circle: I'm still not sure you should synchronize in the destructor at all. Synchronization in CUDA should be explicit.

Reply:
Hmm. Looked at the [documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1gfda584f1788ca983cb21c5f4d2033a62) again:

> In case the device is still doing work in the stream stream when `cudaStreamDestroy()` is called, the function will return immediately and the resources associated with stream will be released automatically once the device has completed all work in stream. 

So, I guess I can just try destroying the stream regardless of anything.

Reply:
Yes!

Reply:
@harrism : Ah, but wait, there's a problem. Right now, the broken way we use streams in API function implementations is:

1. create stream
2. schedule work on stream
3. sync stream
4. destroy stream
5. return

But if we adopt my last suggestion - the stream will not be synchronized when the function returns, i.e. the results will not be ready on the GPU - which means the output is not ready / not usable. In a non-broken system this is fine, because further work will either sync the stream itself or wait on an event or what-not - but again, we don't run that way right now.

So, unfortunately, I have to go back to suggestion we synchronize on orderly destruction to mimic existing behavior. And - I'll add a comment near that statement saying that this is not a good idea and that should be dropped when the use of streams is reworked later on.

The alternative would be for every API function impl to have to explicitly synchronize - but part of the motivation for this wrapper is avoiding just that.

Reply:
@harrism : I could add a boolean template parameter for whether or not to synch  on destruction, I guess.

Reply:
I really don't follow. If the stream is not passed into the function, then it is created in the function. In that case, the function should sync it explicitly before the stream is deleted. Otherwise, why create a stream?

Also, what happens if I pass a scoped_stream by value as a parameter to a function call? That creates a copy, which will be destroyed on exit from the function, which would destroy the stream, which is an oops.

Reply:
1. I wrote this class to correspond to the existing coding pattern I mentioned above. It reduces repetition of code over many API functions (DRY principle), and ensures we don't forget to synch and destroy the stream. We just define a scoped stream variable and forget about stream management.
2. You can't pass a scoped_stream by value - there's no copy constructor. But - I should explicitly delete the copy ctor and assignment operator. When we start working with streams properly, we won't be passing this class around - we would use it at most in the uppermost level. It's like std::unique_ptr in this respect - you have it in your outermost function but pass the raw ptr to functions which use it without deallocating.

Reply:
Can you please link to the pattern in use within libcudf (preferably in multiple places)?

Reply:
If you just search for `stream` under `cpp/src` you'll find a few:

* `binops.cpp`, `binary_valid_mask_and()`
* `bitmask_ops.cu`, `gdf_count_nonzero_mask()`,
* `datetime_ops.cu`, `gdf_extract_datetime_year()`

etc. Linkification coming soon.

Reply:
OK, so I suppose it doesn't hurt to synchronize an already synchronized stream. And it doesn't really hurt to synchronize an already destroyed stream. So I'm back to the code I suggested:

```
~scoped_stream() {
  if (not std::uncaught_exception()) {
    if (cudaSuccess != cudaStreamSynchronize(stream_))
      cudaStreamDestroy(stream_);
  }
}
```

This 
 - doesn't synchronize in the face of an uncaught exceptions
 - doesn't destroy if the stream doesn't synchronize

Reply:
>*    doesn't synchronize in the face of an uncaught exceptions
That's good.

>*  doesn't raise exceptions

Also good.

>*  doesn't destroy if the stream doesn't synchronize

That means the stream will leak. But - what the hell, let's do that. CUDA having a single global status is already a bit silly, we might as well do things this way.

Reply:
The != above should have been a ==...  but you get the point. :)

If we are worried about leaks, just attempt cudaStreamDestroy on all results of cudaStreamSynchronize *other* than cudaStreamInvalidResource.